# Data Science Bootcamp Week1
---

### Our Goal
---
- Analyze the data to be able to understand the salaries of Data Analysts 
- This will enable Data Analysts to be able to the expected salary in respect to various variables at hand


## Exploratory Data Analysis (EDA)

Exploratory Data Analysis is a process used to analyse the datasets and summarize their main characteristics

It helps look at data before making any assumptions.
It helps understand patterns within the data, detect outliers or anomalous events.

<strong> Importing Libraries </strong><br/>
- pandas 
- numpy 
- matplotlib - it a library used for plotting visualizations in python.

In [51]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


## Loading DataSets

In [52]:
it_salary_2018 = pd.read_csv('archive/IT Salary Survey EU 2018.csv')
it_salary_2019 = pd.read_csv('archive/IT Salary Survey EU 2019.csv')
it_salary_2020 = pd.read_csv('archive/IT Salary Survey EU  2020.csv')

Joining the data from the three years:<br/>
- 2018
- 2019
- 2020

In [53]:
it_salary = pd.concat([it_salary_2018, it_salary_2019, it_salary_2020], ignore_index=True)

The consolidated data overview

In [54]:
it_salary.head(3)

Timestamp   Age Gender     City                       Position  \
0  14/12/2018 12:41:33  43.0      M  München                   QA Ingenieur   
1  14/12/2018 12:42:09  33.0      F  München  Senior PHP Magento developer    
2  14/12/2018 12:47:36  32.0      M  München              Software Engineer   

   Years of experience Your level  Current Salary  Salary one year ago  \
0                 11.0     Senior         77000.0              76200.0   
1                  8.0     Senior         65000.0              55000.0   
2                 10.0     Senior         88000.0              73000.0   

   Salary two years ago  ... Years of experience in Germany  \
0               68000.0  ...                            NaN   
1               55000.0  ...                            NaN   
2               54000.0  ...                            NaN   

  Other technologies/programming languages you use often  \
0                                                NaN       
1                                                NaN       
2                                                NaN       

  Yearly brutto salary (without bonus and stocks) in EUR  \
0                                                NaN       
1                                                NaN       
2                                                NaN       

  Yearly bonus + stocks in EUR  \
0                          NaN   
1                          NaN   
2                          NaN   

  Annual brutto salary (without bonus and stocks) one year ago. Only answer if staying in the same country  \
0                                                NaN                                                         
1                                                NaN                                                         
2                                                NaN                                                         

  Annual bonus+stocks one year ago. Only answer if staying in same country  \
0                                                NaN                         
1                                                NaN                         
2                                                NaN                         

  Employment status Have you lost your job due to the coronavirus outbreak?  \
0               NaN                                                NaN        
1               NaN                                                NaN        
2               NaN                                                NaN        

   Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week  \
0                                                NaN                                                   
1                                                NaN                                                   
2                                                NaN                                                   

   Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR  
0                                                NaN                                                                        
1                                                NaN                                                                        
2                                                NaN                                                                        

[3 rows x 42 columns]

## Understanding the Data
---

In [55]:
it_salary.shape

(3009, 42)

In [56]:
it_salary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3009 entries, 0 to 3008
Data columns (total 42 columns):
 #   Column                                                                                                                   Non-Null Count  Dtype  
---  ------                                                                                                                   --------------  -----  
 0   Timestamp                                                                                                                2018 non-null   object 
 1   Age                                                                                                                      2780 non-null   float64
 2   Gender                                                                                                                   2985 non-null   object 
 3   City                                                                                                                     2980 non-null   o

In [57]:
it_salary.notnull().sum()

Timestamp                                                                                                                  2018
Age                                                                                                                        2780
Gender                                                                                                                     2985
City                                                                                                                       2980
Position                                                                                                                    737
Years of experience                                                                                                        1723
Your level                                                                                                                  743
Current Salary                                                                                          

The DataFrame contains:
- 3009 entries
- 42 features
- Non of the features is complete 

Some Columns are not instrumental in determining the salary:
- `Have you lost your job due to the coronavirus outbreak?`
- `Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week:`
- `Have you received additional monetary support from your employer due to Work From Home? If yes, how much 2020 in EUR:`

Drop the columns

In [58]:
it_salary.columns

Index(['Timestamp', 'Age', 'Gender', 'City', 'Position', 'Years of experience',
       'Your level', 'Current Salary', 'Salary one year ago',
       'Salary two years ago', 'Are you getting any Stock Options?',
       'Main language at work', 'Company size', 'Company type', 'Zeitstempel',
       'Seniority level', 'Position (without seniority)',
       'Your main technology / programming language',
       'Yearly brutto salary (without bonus and stocks)', 'Yearly bonus',
       'Yearly stocks',
       'Yearly brutto salary (without bonus and stocks) one year ago. Only answer if staying in same country',
       'Yearly bonus one year ago. Only answer if staying in same country',
       'Yearly stocks one year ago. Only answer if staying in same country',
       'Number of vacation days', 'Number of home office days per month',
       'Company name ', 'Сontract duration', 'Company business sector', '0',
       'Position ', 'Total years of experience',
       'Years of experience in Germa

In [59]:
it_salary = it_salary.drop('Annual brutto salary (without bonus and stocks) one year ago. Only answer if staying in the same country', axis=1)
it_salary = it_salary.drop('Have you been forced to have a shorter working week (Kurzarbeit)? If yes, how many hours per week', axis=1)
it_salary = it_salary.drop('Have you received additional monetary support from your employer due to Work From Home? If yes, how much in 2020 in EUR', axis=1)
it_salary = it_salary.drop('Timestamp', axis=1)
it_salary = it_salary.drop('Yearly bonus + stocks in EUR', axis=1)
it_salary = it_salary.drop('Annual bonus+stocks one year ago. Only answer if staying in same country', axis=1)
it_salary.shape

(3009, 36)

<strong>Rename Columns for Readability</strong>


In [60]:
it_salary.rename(columns = {
    'Total years of experience':'experience',
    'Years of experience in Germany':'germany_experience',
    'Seniority level':'work_level',
    'Your main technology / programming language':'main_tech',
    'Other technologies/programming languages you use often':'other_tech',
    'Number of vacation days':'vacation_days',
    'Yearly brutto salary (without bonus and stocks) in EUR':'yearly_salary',
    'Employment status':'emp_state',
    'Сontract duration':'contract_duration',
    'Main language at work':'main_language',
    'Company size':'company_size',
    'Company type':'company_type',
    'Have you lost your job due to the coronavirus outbreak?':'covid_lost',
    'Age':'age',
    'Gender':'gender',
    'City':'city',
    'Position ':'position'
    
}, inplace = True)

In [61]:
it_salary

age gender         city                       Position  \
0     43.0      M      München                   QA Ingenieur   
1     33.0      F      München  Senior PHP Magento developer    
2     32.0      M      München              Software Engineer   
3     25.0      M      München      Senior Frontend Developer   
4     39.0      M      München                    UX Designer   
...    ...    ...          ...                            ...   
3004  31.0   Male       Berlin                            NaN   
3005  33.0   Male       Berlin                            NaN   
3006  39.0   Male       Munich                            NaN   
3007  26.0   Male  Saarbrücken                            NaN   
3008  26.0   Male       Berlin                            NaN   

      Years of experience Your level  Current Salary  Salary one year ago  \
0                    11.0     Senior         77000.0              76200.0   
1                     8.0     Senior         65000.0              55000.0   
2                    10.0     Senior         88000.0              73000.0   
3                     6.0     Senior         78000.0              55000.0   
4                    10.0     Senior         69000.0              60000.0   
...                   ...        ...             ...                  ...   
3004                  NaN        NaN             NaN                  NaN   
3005                  NaN        NaN             NaN                  NaN   
3006                  NaN        NaN             NaN                  NaN   
3007                  NaN        NaN             NaN                  NaN   
3008                  NaN        NaN             NaN                  NaN   

      Salary two years ago Are you getting any Stock Options?  ...  \
0                  68000.0                                 No  ...   
1                  55000.0                                 No  ...   
2                  54000.0                                 No  ...   
3                  45000.0                                Yes  ...   
4                  52000.0                                 No  ...   
...                    ...                                ...  ...   
3004                   NaN                                NaN  ...   
3005                   NaN                                NaN  ...   
3006                   NaN                                NaN  ...   
3007                   NaN                                NaN  ...   
3008                   NaN                                NaN  ...   

       contract_duration Company business sector   0  \
0                    NaN                     NaN NaN   
1                    NaN                     NaN NaN   
2                    NaN                     NaN NaN   
3                    NaN                     NaN NaN   
4                    NaN                     NaN NaN   
...                  ...                     ...  ..   
3004  Unlimited contract                     NaN NaN   
3005  Unlimited contract                     NaN NaN   
3006  Unlimited contract                     NaN NaN   
3007  Unlimited contract                     NaN NaN   
3008  Unlimited contract                     NaN NaN   

                                    position experience germany_experience  \
0                                        NaN        NaN                NaN   
1                                        NaN        NaN                NaN   
2                                        NaN        NaN                NaN   
3                                        NaN        NaN                NaN   
4                                        NaN        NaN                NaN   
...                                      ...        ...                ...   
3004                       Backend Developer          9                  5   
3005  Researcher/ Consumer Insights Analyst          10                1.5   
3006                   IT Operations Manager         15                  2   
3007   

<strong>Check Structure Error and Null Values</strong>

In [62]:
for i in it_salary.columns:
    print(i, ': Null Values are ', it_salary[i].isnull().sum(),
          '\n', it_salary[i].unique(),
          '\n', '-'*50)

age : Null Values are  229 
 [43. 33. 32. 25. 39. 37. nan 36. 30. 40. 31. 29. 24. 44. 34. 28. 35. 38.
 27. 41. 26. 22. 42. 52. 50. 49. 23. 21. 48. 45. 60. 47. 46. 51. 54. 20.
 59. 66. 56. 53. 65. 69.] 
 --------------------------------------------------
gender : Null Values are  24 
 ['M' 'F' nan 'Male' 'Female' 'Diverse'] 
 --------------------------------------------------
city : Null Values are  29 
 ['München' 'Berlin' 'Karlsruhe' 'Hamburg' 'Augsburg ' 'Bremen' 'Frankfurt'
 nan 'Regensburg' 'Lisbon' 'Leipzig' 'Stuttgart' 'Amsterdam' 'Luxembourg'
 'Bern' 'Bonn' 'Köln' 'Baden-Württemberg ' 'Frankonia' 'Stuttgart '
 'Mannheim ' 'Bayern' 'Tübingen' 'Brno' 'Dresden' 'Krakow' 'Dublin'
 'Eindhoven' 'Amsterdam ' 'Konstanz' 'Dublin ' 'Prague' 'Moscow'
 'Stockholm' 'Belgrade, Serbia' 'Kyiv' 'Nürnberg' 'Karlskrona' 'Krakau'
 'Vienna' 'London' 'Bitburg' 'Gdansk' 'Poland' 'Düsseldorf' 'Ludwigsburg'
 'Essen' 'Strasbourg' 'Alicante, Spain' 'Warsaw' 'Memmingen' 'Tallinn'
 'Zurich' 'oslo' 'Münster'

In [63]:
it_salary.describe()

age  Years of experience  Current Salary  Salary one year ago  \
count  2780.000000          1723.000000      750.000000           596.000000   
mean     32.401079             8.537725    68381.765333         62187.278523   
std       5.406931             4.906948    21196.306557         20163.008663   
min      20.000000             0.000000    10300.000000         10001.000000   
25%      29.000000             5.000000    57000.000000         52000.000000   
50%      32.000000             8.000000    65000.000000         60000.000000   
75%      35.000000            11.000000    75000.000000         70000.000000   
max      69.000000            38.000000   200000.000000        200000.000000   

       Salary two years ago  Yearly brutto salary (without bonus and stocks)  \
count            463.000000                                       990.000000   
mean           58013.475162                                     72562.212121   
std            20413.048908                                     23947.379182   
min            10001.000000                                      6000.000000   
25%            48000.000000                                     60000.000000   
50%            56000.000000                                     70000.000000   
75%            67000.000000                                     80000.000000   
max           150000.000000                                    216000.000000   

       Yearly bonus  Yearly stocks  \
count    530.000000     203.000000   
mean    7857.169811   18263.119803   
std    10320.628344   61012.387773   
min        0.000000       0.000000   
25%      300.000000       1.000000   
50%     5000.000000    2000.000000   
75%    10000.000000   20000.000000   
max    80000.000000  750000.000000   

       Yearly brutto salary (without bonus and stocks) one year ago. Only answer if staying in same country  \
count                                         603.000000                                                      
mean                                        65803.895522                                                      
std                                         20554.017614                                                      
min                                             0.000000                                                      
25%                                         55000.000000                                                      
50%                                         65000.000000                                                      
75%                                         75000.000000                                                      
max                                        200000.000000                                                      

       Yearly bonus one year ago. Only answer if staying in same country  \
count                                         257.000000                   
mean                                         8144.000000                   
std                                         15067.876701                   
min                                             0.000000                   
25%                                          1000.000000                   
50%                                          5000.000000                   
75%                                          9000.000000                   
max                                        150000.000000                   

       Yearly stocks one year ago. Only answer if staying in same country  \
count                                         139.000000                    
mean                                         8204.476331                    
std                                         45224.397374                    
min                                             0.000000                    
25%                                             0.000000                    
50%                                             0.100000                    
75%   

### Filling Null Values

In [69]:
it_salary['age'] = it_salary['age'].fillna(it_salary.groupby('experience')['age'].transform('mean'))


#drop rows with null values and null experience
it_salary.dropna(subset=['age'], how='all', inplace=True)

#Changing the age datatype to int
